core/consensus/xpoa/xpoa.go

```diff
@@ -351,8 +351,14 @@ Again:
 // getCurrentValidates return current validates from xmodel
 // 注意：当查不到的时候或者一个候选人都没查到则默认取初始化的值
 // TODO: zq needs to be optimized in future because
-func (xpoa *XPoa) getCurrentValidates() ([]*cons_base.CandidateInfo, int64, int64, error) {
-	contractRes, confirmedTime, confirmedHeight, err := xpoa.utxoVM.SystemCall(xpoa.xpoaConf.contractName, xpoa.xpoaConf.methodName, nil, true)
+func (xpoa *XPoa) getCurrentValidates(curHeight int64) ([]*cons_base.CandidateInfo, int64, int64, error) {
+	preBlockId, err := xpoa.ledger.QueryBlockByHeight(curHeight)
+	xmod := xpoa.utxoVM.GetXModel()
```

Comment:
May be it's better to package this process in SysCall function. Or utxovm return an snapshot reader.

Reply:
SystemCall is a general function with different XMReader, say, XModel or SnapShot, for preparing model cache env to invoke contract. In my opinion, preparing XMReader outside the SystenCall function might a better way in view of further general usage.

Reply:
Utxovm return an snapshot reader: done :)